### PR TITLE
CP-24132: unit test for usb_scan.py

### PR DESCRIPTION
**Pull request by TalonsLee**
add unit test for usb_scan.py

Signed-off-by: Xin(Talons) Li <xin.li@citrix.com>

### CODE REVIEW

File: scripts/test_usb_scan.py [hunk 1]
```diff
@@ -0,0 +1,393 @@
...
+def mock_setup(mod, devices, interfaces, path):
+    mod.log.error = test_log
+    mod.log.debug = test_log
+    mod.Policy._PATH = path
+    mod.pyudev.Context = mock.Mock(return_value=MocContext(
+        devices, interfaces))
+
+
+@nottest
+def test_log(m):
```
Comment (mseri) [inline scripts/test_usb_scan.py]: I would avoid using a file, instead I'd mock the stdout or the stderr and use print.
An example is
```python
>>> from StringIO import StringIO
>>> def foo():
...     print 'Something'
...
>>> @patch('sys.stdout', new_callable=StringIO)
... def test(mock_stdout):
...     foo()
...     assert mock_stdout.getvalue() == 'Something\n'
...
>>> test()
```
from the mock documentation at http://www.voidspace.org.uk/python/mock/patch.html

However it doesn't seem that you are using the logs content at all. In that case just replace the log funciton with a void function

File: scripts/test_usb_scan.py [hunk 2]
```diff
@@ -0,0 +1,393 @@
...
+        pusbs = [usb_scan.to_pusb(d) for d in devices if
+                 d.is_ready() and policy.check(d)]
+
+        # pass pusbs in json to XAPI
+        self.assertItemsEqual(pusbs, moc_results)
+
+    @nottest
+    def test_usb_exit(self, moc_devices, moc_interfaces, moc_results,
+                        path="./scripts/usb-policy.conf"):
+        with self.assertRaises(SystemExit):
```
Comment (mseri) [inline scripts/test_usb_scan.py]: I think here you should also check that the failure did show the correct log error, at least the immutable part of it.
You could use `@raises(SystemExit)` as attribute and `assert` over the initial part of the error message for example

File: scripts/test_usb_scan.py [hunk 3]
```diff
@@ -0,0 +1,393 @@
...
+
+    @nottest
+    def test_usb_common(self, moc_devices, moc_interfaces, moc_results,
+                        path="./scripts/usb-policy.conf"):
+        import usb_scan
+        mock_setup(usb_scan, moc_devices, moc_interfaces, path)
+
+        devices, interfaces = usb_scan.get_usb_info()
+
+        # debug info
```
Comment (mseri) [inline scripts/test_usb_scan.py]: We should not copy paste the code, instead we should refactor `usb_scan.py` to be better suited for testing.
I think we need to extract https://github.com/xapi-project/xen-api/blob/3bb747104ee0399c3d0e3c747b1d8ff95669b269/scripts/usb_scan.py#L627-L637 into a function that takes `devices` and `interfaces` as parameters and returns the object that we will serialise to json. Then this test should call that function with the mocked devices and interfaces

File: scripts/test_usb_scan.py [hunk 4]
```diff
@@ -0,0 +1,393 @@
...
+        self.test_usb_exit([], [], [], "not_exit.conf")
+
+    @nottest
+    def test_usb_config_error_common(self, content):
+        path = os.path.join(self.work_dir, "usb-policy.conf")
+        with open(path, "w") as f:
+            f.write(content)
+        self.test_usb_exit([], [], [], path)
+
+    def test_usb_config_error_01(self):
```
Comment (mseri) [inline scripts/test_usb_scan.py]: This is just a stylistic thing. Here and below, give meaningful names to the error functions. For example, in this case, `test_usb_config_error_unexpected_characters` and in the next one `test_usb_config_error_duplicated_key`, and so on

File: scripts/test_usb_scan.py (hunk 1 above)
Comment (TalonsLee) [reply scripts/test_usb_scan.py]: OK. remove log file, and make test_log empty.

File: scripts/test_usb_scan.py (hunk 3 above)
Comment (TalonsLee) [reply scripts/test_usb_scan.py]: OK. This will change the usb_scan.py,
I have to run more test next week.

File: scripts/test_usb_scan.py (hunk 4 above)
Comment (TalonsLee) [reply scripts/test_usb_scan.py]: done.

File: scripts/test_usb_scan.py (hunk 2 above)
Comment (TalonsLee) [reply scripts/test_usb_scan.py]: OK.
    @nottest
    def test_usb_exit(self, devices, interfaces, results,
                      path="./scripts/usb-policy.conf", msg=""):
        with self.assertRaises(SystemExit) as cm:
            self.test_usb_common(devices, interfaces, results, path)
        exception = cm.exception
        print exception.message
        if msg:
            self.assertIn(msg, exception.message)

File: scripts/test_usb_scan.py (hunk 3 above)
Comment (mseri) [reply scripts/test_usb_scan.py]: Yes, call it something like `make_pusbs_list(interfaces, devices)` and the unit test that function

File: scripts/usb_scan.py [hunk 5]
```diff
@@ -607,19 +607,14 @@ def to_pusb(device):
...
+def make_pusbs_list(devices, interfaces):
+    """ check the USB devices and interfaces against policy file,
+    and return the pusb list that can be passed through
 
+    :param devices:([UsbDevice]) USB device list we found in host
+    :param interfaces:([UsbInterface]) USB interface list found in host
+    :return:([pusb]) the pusb list that can be passed through
+    """
     # debug info
     log_list(devices)
```
Comment (mseri) [inline scripts/usb_scan.py]: Please move these debug info logs to the main body before L648, or remove them completely

**Pull request merged**